src/uct/cuda/cuda_copy/cuda_copy_ep.c

```diff
@@ -41,25 +43,59 @@ UCS_CLASS_DEFINE_DELETE_FUNC(uct_cuda_copy_ep_t, uct_ep_t);
      ucs_trace_data(_fmt " to %"PRIx64"(%+ld)", ## __VA_ARGS__, (_remote_addr), \
                     (_rkey))
 
-#define UCT_CUDA_COPY_CHECK_AND_CREATE_STREAM(_iface, _id) \
-    if ((_iface)->stream[_id] == 0) { \
+/* It is relatively inexpensive to query memory attributes compared to actual
+ * cudaMemcpy cost. Query time is between 0.05 to 0.2 us on x86 but memcpy cost
+ * is generally > 7 us */
```

Comment:
2 detection calls for each fragment, both at sender and receiver.  this is the considerable cost when running collectives.
unless, if there is real advantage to differentiate stream for cuda and managed, maybe we can stick to existing stream per direction

Reply:
Approximately 2% overhead per cudaMemcpy on x86

Reply:
the % might be high for D2D, right?
one recent observation w/o memtype is that in ~1k alltoall at scale(128 GPUs) caused ~100 us overhead because it had to detect the memory type for each target send. so, just critical about if we really need to differentiate the stream for cuda and managed and have this detection in here

Reply:
Cost of D2D is slightly better than D2H, H2D for short/medium messages but it starts at about 6us. That being the case, at 0.2 us overhead per cudaMemcpy call should still be reasonable. 

> one recent observation w/o memtype is that in ~1k alltoall at scale(128 GPUs) caused ~100 us overhead because it had to detect the memory type for each target send. 

~100us out of how much? What's the % overhead added?

Reply:
It is a user reported. collective time is around 500- 600 us

Reply:
can you check the impact in osu_bw and osu_bibw wih pipelining ( with rndv thread =16k and GDR off)?

Reply:
osu_bw | PR | master | % diff |   | osu_bibw | PR | master | % diff
-- | -- | -- | -- | -- | -- | -- | -- | --
32768 | 2675.8 | 2804.09 | 4 |   | 32768 | 2998.71 | 3334.12 | 10
65536 | 4159.86 | 4416.6 | 5 |   | 65536 | 5133.93 | 5795.01 | 11
131072 | 5818.88 | 6030.88 | 3 |   | 131072 | 7453.65 | 7703.97 | 3
262144 | 6933.65 | 7062.72 | 1 |   | 262144 | 8734.18 | 8921.18 | 2
524288 | 7602.95 | 7681.72 | 1 |   | 524288 | 9601.15 | 9816.3 | 2
1048576 | 7696.75 | 7765.13 | 0 |   | 1048576 | 9814.6 | 9917.04 | 1
2097152 | 7768.04 | 7827.27 | 0 |   | 2097152 | 9914.29 | 10024.19 | 1
4194304 | 7785.14 | 7844.47 | 0 |   | 4194304 | 9738.23 | 9844.27 | 1

This is on DGX-1. The difference is considerable for 32k and 64k but this seems to be coming from pipeline bw and bibw performance regression in master and 1.10.x compared 1.9.x. Once the perf regression gets resolved, the overhead should again be negligible.

Reply:
Adding perf results of querying two pointers before memcpy here as well. 
[cumemcpy-w-query.xlsx](https://github.com/openucx/ucx/files/5822248/cumemcpy-w-query.xlsx)